h2/src/main/org/h2/index/AbstractFunctionCursor.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2004-2018 H2 Group. Multiple-Licensed under the MPL 2.0,
+ * and the EPL 1.0 (http://h2database.com/html/license.html).
+ * Initial Developer: H2 Group
+ */
+package org.h2.index;
+
+import org.h2.engine.Session;
+import org.h2.message.DbException;
+import org.h2.result.Row;
+import org.h2.result.SearchRow;
+import org.h2.value.Value;
+
+/**
+ * Abstract function cursor.
+ */
+abstract class AbstractFunctionCursor implements Cursor {
+    private final FunctionIndex index;
+
+    private final SearchRow first;
+
+    private final SearchRow last;
+
+    final Session session;
+
+    Value[] values;
+
+    Row row;
+
+    /**
+     * @param index
+     *            index
+     * @param first
+     *            first row
+     * @param last
+     *            last row
+     * @param session
+     *            session
+     */
+    AbstractFunctionCursor(FunctionIndex index, SearchRow first, SearchRow last, Session session) {
+        this.index = index;
+        this.first = first;
+        this.last = last;
+        this.session = session;
+    }
+
+    @Override
+    public Row get() {
+        if (values == null) {
+            return null;
+        }
+        if (row == null) {
+            row = session.createRow(values, 1);
+        }
+        return row;
+    }
+
+    @Override
+    public SearchRow getSearchRow() {
+        return get();
+    }
+
+    @Override
+    public boolean next() {
```

Comment:
not sure I see this point of this? as far as I can tell you've just moved the full scan here instead of the full scan happening at the layer of code above this? what am I missing?

Reply:
Filtration in `AbstractFunctionCursor` now probably can be removed because this issue is fixed in `IndexCursor`, but I'm not sure that `IndexCusror` was the only one caller than expects correct `find()` implementation.